### PR TITLE
CloudWatch: Make it possible to specify custom api endpoint

### LINKED ISSUE #26041: Cloudwatch: Custom endpoint URL in CloudWatch data source
**What would you like to be added**:

In CloudWatch data source, we would like to have possibility to override default AWS endpoint URL with custom value.

**Why is this needed**:

In one of our scenarios, we want to use custom endpoint URL instead of default one (as we already can do in AWS CLI). 
Change should be rather simple: passing additional parameter from data source configuration to SDK on init, and adding optional field in Cloudwtch datasource view in UI.

### PR DESCRIPTION
**Pull request by sunker**
Tested using `monitoring.us-east-2.amazonaws.com` and it works fine. Info about api endpoints and why to use them can be found [here](https://docs.aws.amazon.com/general/latest/gr/rande.html)

All the other AWS data source plugins that we have, i.e IOT Sitewise, Timestream and X-Ray, already have support for specifying a custom endpoint. 

Fixes #26041

### CODE REVIEW

File: docs/sources/datasources/cloudwatch.md [hunk 1]
```diff
@@ -105,6 +105,10 @@ Here is a minimal policy example:
 
 The `Assume Role ARN` field allows you to specify which IAM role to assume, if any. When left blank, the provided credentials are used directly and the associated role or user should have the required permissions. If this field is non-blank, on the other hand, the provided credentials are used to perform an [sts:AssumeRole](https://docs.aws.amazon.com/STS/latest/APIReference/API_AssumeRole.html) call.
 
+### Endpoint
+
+The `Endpoint` field allows you to specify a custom endpoint URL that overrides the default generated endpoint for the CloudWatch API. Leave this field blank to use the default generated endpoint. Read about why and how to use Service endpoints in the [official AWS documentation](https://docs.aws.amazon.com/general/latest/gr/rande.html).
```
Comment (achatterjee-grafana) [inline docs/sources/datasources/cloudwatch.md:110]:
```suggestion
The `Endpoint` field allows you to specify a custom endpoint URL that overrides the default generated endpoint for the CloudWatch API. Leave this field blank if you want to use the default generated endpoint. For more information on why and how to use Service endpoints in the [official AWS documentation](https://docs.aws.amazon.com/general/latest/gr/rande.html).
```

File: public/app/plugins/datasource/cloudwatch/components/ConfigEditor.tsx [hunk 2]
```diff
@@ -314,6 +314,21 @@ export class ConfigEditor extends PureComponent<Props, State> {
               />
             </div>
           </div>
+          <div className="gf-form-inline">
+            <div className="gf-form">
+              <InlineFormLabel className="width-14" tooltip="Optionally, specify a custom endpoint for the service">
```
Comment (achatterjee-grafana) [inline public/app/plugins/datasource/cloudwatch/components/ConfigEditor.tsx:319]:
```suggestion
              <InlineFormLabel className="width-14" tooltip="Optionally, specify a custom endpoint for the service.">
```

**Review by achatterjee-grafana — CHANGES_REQUESTED** [review]
Added minor copy-edit suggestions.

File: pkg/tsdb/cloudwatch/cloudwatch.go [hunk 3]
```diff
@@ -129,6 +130,10 @@ func (e *cloudWatchExecutor) newSession(region string) (*session.Session, error)
 		regionCfg = &aws.Config{Region: aws.String(dsInfo.Region)}
 		cfgs = append(cfgs, regionCfg)
 	}
+	
+	if dsInfo.Endpoint != "" {
+		cfgs = append(cfgs, &aws.Config{ Endpoint: aws.String(dsInfo.Endpoint)})
```
[the remark this answers is not in the document]
Comment (sunker) [reply pkg/tsdb/cloudwatch/cloudwatch.go]: Yes!

File: docs/sources/datasources/cloudwatch.md [hunk 4]
```diff
@@ -105,6 +105,10 @@ Here is a minimal policy example:
 
 The `Assume Role ARN` field allows you to specify which IAM role to assume, if any. When left blank, the provided credentials are used directly and the associated role or user should have the required permissions. If this field is non-blank, on the other hand, the provided credentials are used to perform an [sts:AssumeRole](https://docs.aws.amazon.com/STS/latest/APIReference/API_AssumeRole.html) call.
 
+### Endpoint
+
+The `Endpoint` field allows you to specify a custom endpoint URL that overrides the default generated endpoint for the CloudWatch API. Leave this field blank if you want to use the default generated endpoint. For more information on why and how to use Service endpoints in the [official AWS documentation](https://docs.aws.amazon.com/general/latest/gr/rande.html).
```
[the remark this answers is not in the document]
Comment (sunker) [reply docs/sources/datasources/cloudwatch.md:110]: Oh yes, good catch.

**Review by aknuds1 — APPROVED** [review]

File: docs/sources/datasources/cloudwatch.md [hunk 5]
```diff
@@ -105,6 +105,10 @@ Here is a minimal policy example:
 
 The `Assume Role ARN` field allows you to specify which IAM role to assume, if any. When left blank, the provided credentials are used directly and the associated role or user should have the required permissions. If this field is non-blank, on the other hand, the provided credentials are used to perform an [sts:AssumeRole](https://docs.aws.amazon.com/STS/latest/APIReference/API_AssumeRole.html) call.
 
+### Endpoint
+
+The `Endpoint` field allows you to specify a custom endpoint URL that overrides the default generated endpoint for the CloudWatch API. Leave this field blank if you want to use the default generated endpoint. For more information on why and how to use Service endpoints, see the [official AWS documentation](https://docs.aws.amazon.com/general/latest/gr/rande.html).
```
Comment (osg-grafana) [inline docs/sources/datasources/cloudwatch.md:110]:
```suggestion
The `Endpoint` field allows you to specify a custom endpoint URL that overrides the default generated endpoint for the CloudWatch API. Leave this field blank if you want to use the default generated endpoint. For more information on why and how to use Service endpoints, refer to the [AWS service endpoints](https://docs.aws.amazon.com/general/latest/gr/rande.html).
```
Conform to style guide, which I am about to update to include this advice.

**Review by osg-grafana — CHANGES_REQUESTED** [review]

File: docs/sources/datasources/cloudwatch.md (hunk 5 above)
Comment (aknuds1) [reply docs/sources/datasources/cloudwatch.md:110]: Shouldn't it instead be something like "the AWS service endpoints documentation"? "refer to the AWS service endpoints." doesn't make sense to me language wise.

Comment (sunker) [reply docs/sources/datasources/cloudwatch.md]: I agree. Text updated

Comment (osg-grafana) [reply docs/sources/datasources/cloudwatch.md]: Typically, we default to WYSIWYG when referring out to documentation. Try to reword such that you can use the verbatim title of the referred to document.

Comment (aknuds1) [reply docs/sources/datasources/cloudwatch.md]: It reads fine as it is though.

**Review by osg-grafana — APPROVED** [review]
Unblocking (if I was blocking this PR) with an Approve

**Review by achatterjee-grafana — APPROVED** [review]

**Comment by achatterjee-grafana** [comment]
Good catch, "documentation" word was missing and it is needed here.

**Comment by bitbier** [comment]
I actually commented on the wrong PR, but I don't think this will work for both metrics and logs or there should be a documented caveat on that. 

Cloudwatch Metrics and Cloudwatch Logs both use different endpoints `monitoring.us-east-2.amazonaws.com` and `logs.us-east-2.amazonaws.com` respectively; however, the datasource mixes both metrics and logs into one datasource. I think the current workaround here is to create two separate CloudWatch datasources (one for metrics and one for logs) if you want override the endpoint here.

**Comment by yarons** [comment]
We ran into this problem as well but adding to different data sourced still leads to an error and the data is not fetched.
Described in #63420.

**Pull request merged**